src/main.rs

```diff
@@ -0,0 +1,423 @@
+        Err(tonic::Status::unimplemented("endpoint not implemented"))
+    }
+
+    async fn node_unstage_volume(
+        &self,
+        _request: Request<NodeUnstageVolumeRequest>,
+    ) -> Result<Response<NodeUnstageVolumeResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented("endpoint not implemented"))
+    }
+
+    async fn node_publish_volume(
+        &self,
+        request: Request<NodePublishVolumeRequest>,
+    ) -> Result<Response<NodePublishVolumeResponse>, tonic::Status> {
+        let request = request.into_inner();
+        let target_path = PathBuf::from(request.target_path);
+        tracing::info!(
+            volume.path = %target_path.display(),
+            volume.ctx = ?request.volume_context,
+            "Received NodePublishVolume request"
+        );
+        let sel = SecretVolumeSelector::deserialize(request.volume_context.into_deserializer())
+            .map_err(PublishError::InvalidSelector)?;
+        let data = self.get_secret_data(sel).await?;
+        self.save_secret_data(&target_path, data).await?;
+        Ok(Response::new(NodePublishVolumeResponse {}))
+    }
+
+    async fn node_unpublish_volume(
+        &self,
+        request: Request<NodeUnpublishVolumeRequest>,
+    ) -> Result<Response<NodeUnpublishVolumeResponse>, tonic::Status> {
+        let request = request.into_inner();
+        let target_path = PathBuf::from(request.target_path);
+        tracing::info!(
+            volume.path = %target_path.display(),
+            "Received NodeUnpublishVolume request"
+        );
+        tokio::fs::remove_dir_all(&target_path)
+            .await
+            .map_err(|err| UnpublishError::Cleanup(err, target_path))?;
+        Ok(Response::new(NodeUnpublishVolumeResponse {}))
+    }
+
+    async fn node_get_volume_stats(
+        &self,
+        _request: Request<NodeGetVolumeStatsRequest>,
+    ) -> Result<Response<NodeGetVolumeStatsResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented("endpoint not implemented"))
+    }
+
+    async fn node_expand_volume(
+        &self,
+        _request: Request<NodeExpandVolumeRequest>,
+    ) -> Result<Response<NodeExpandVolumeResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented("endpoint not implemented"))
+    }
+
+    async fn node_get_capabilities(
+        &self,
+        _request: Request<NodeGetCapabilitiesRequest>,
+    ) -> Result<Response<NodeGetCapabilitiesResponse>, tonic::Status> {
+        Ok(Response::new(NodeGetCapabilitiesResponse {
+            capabilities: vec![],
+        }))
+    }
+
+    async fn node_get_info(
+        &self,
+        _request: Request<NodeGetInfoRequest>,
+    ) -> Result<Response<NodeGetInfoResponse>, tonic::Status> {
+        Ok(Response::new(NodeGetInfoResponse {
+            node_id: "asdf".to_string(),
+            max_volumes_per_node: i64::MAX,
+            accessible_topology: None,
+        }))
+    }
```

Comment:
Why return "fake" values here (asdf, empty) instead of errors?

Reply:
These are required by the CSI API. Only `node_id` is fake here (and we need to provide *something*, otherwise K8s ends up very happy).

Reply:
Ok